tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -986,22 +987,41 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
   SwiftLangSupport::printDeclTypeUSR(DInfo.VD, OS);
   Symbol.TypeUSR = copyAndClearString(Allocator, Buffer);
 
-  // Serialize the SubstitutionMap into an array of string pairs
+  // Serialize the SubstitutionMap into an array of SubstitutionInfo
   if (DInfo.IsRef) {
     PrintOptions Options;
     Options.PrintTypeAliasUnderlyingType = true;
 
-    // Given types T1 and T2, returns "<typename(T1)> : <typename(T2)>" string
-    auto getPrintedPairOfTypes = [&](Type Ty1, Type Ty2) {
-      Ty1->print(OS, Options);
-      OS << " : ";
-      Ty2->print(OS, Options);
-      return copyAndClearString(Allocator, Buffer);
+    // Given generic type and its corresponding replacement type, returns their
+    // SubstitutionInfo
+    auto getSubstitutionInfo = [&](Type GenericTy, Type Ty) {
+      if (!GenericTy || !Ty) {
+        return SubstitutionInfo();
+      }
+      SubstitutionInfo SubstInfo;
+      GenericTy->print(OS, Options);
+      OS << " -> ";
+      Ty->print(OS, Options);
+      SubstInfo.Mapping = copyAndClearString(Allocator, Buffer);
+      if (!ExpandSubstitutions) {
+        return SubstInfo;
+      }
+      auto *Decl1 = GenericTy->getAs<GenericTypeParamType>()->getDecl();
+      auto *Decl2 = Ty->getAnyNominal();
+      if (Decl1) {
+        SwiftLangSupport::printUSR(Decl1, OS);
+        SubstInfo.GenericUSR = copyAndClearString(Allocator, Buffer);
+      }
+      if (Decl2) {
+        SwiftLangSupport::printUSR(Decl2, OS);
+        SubstInfo.ReplacementUSR = copyAndClearString(Allocator, Buffer);
+      }
+      return SubstInfo;
     };
 
     QuerySubstitutionMap GetType{DInfo.Substitutions};
     auto Params = DInfo.Substitutions.getGenericSignature().getGenericParams();
-    SmallVector<StringRef, 2> Substitutions;
+    SmallVector<SubstitutionInfo, 2> Substitutions;
 
     for (auto *GenericTy : Params) {
       // Lookup the replacement type
```

Comment:
You can also iterate over DInfo.Substitutions.getReplacementTypes() instead of looking them up one by one